Cygwin is POSIX, don't say it isn't

levitte opened the pull request:
More to the point, Cygwin is a POSIX API.  In our library, the use of
a POSIX API is marked by defining the macro OPENSSL_SYS_UNIX.
Therefore, that macro shouldn't be undefined when building for Cygwin.

levitte committed:
Cygwin is POSIX, don't say it isn't

More to the point, Cygwin is a POSIX API.  In our library, the use of
a POSIX API is marked by defining the macro OPENSSL_SYS_UNIX.
Therefore, that macro shouldn't be undefined when building for Cygwin.

levitte commented:
Without this change, we get non-function `OPENSSL_DIR_*` functions, i.e. the "not yet implemented" implementation, and because of that, `test/recipes/90-test_store.t` fails.

It builds just as cleanly, and tests just as cleanly after this change.

t-j-h commented:
Making this change alters more than the directory handling routines.
For example: Does cygwin have a fully functional mmap?  secmem routines use that for unix platforms ...

levitte commented:
Like I said, it builds fine and it tests fine.  That includes `test/recipes/90-test_secmem.t`, which runs `test/secmemtest`.  Just now, I ran that one in the debugger, just to see how `sh_init` would behave, and sure enough, it did the `mmap` call with no errors.

levitte commented:
Sure.  I'll change the commit message

levitte commented:
Commit message changed, and I've adapted the PR title and description accordingly

dot-asm commented:
Unfortunately I'm not in position to verify this for the moment, and won't be for several more days. So if I'm to approve this, I can't do it only when I'm back home... Sorry :-(

dot-asm commented:
Just in case, it looks appropriate, but I'd prefer to verify [even with older Cygwin I have lying around].

levitte commented:
Sure.  I'm in no rush

bernd-edlinger commented:
there are a few places which could be cleaned up once OPENSSL_SYS_UNIX and
OPENSSL_SYS_CYGWIN are no longer alternatives:
```
crypto/async/arch/async_posix.h:#if (defined(OPENSSL_SYS_UNIX) || defined(OPENSSL_SYS_CYGWIN)) \
engines/e_dasync.c:#if (defined(OPENSSL_SYS_UNIX) || defined(OPENSSL_SYS_CYGWIN)) && defined(OPENSSL_THREADS)
```

levitte commented:
Goot catch!

levitte committed:
Simplify Cygwin checks, part 1

Because OPENSSL_SYS_CYGWIN will keep OPENSSL_SYS_UNIX defined, there's
no point having checks of this form:

    #if (defined(OPENSSL_SYS_UNIX) || defined(OPENSSL_SYS_CYGWIN))

levitte committed:
The Cygwin gcc doesn't define _WIN32, don't pretend it does

levitte commented:
I added a couple more commits to clean up the checks where `OPENSSL_SYS_CYGWIN` and `__CYGWIN__` do not need to be checked (any more for the former, and was never necessary for the latter).

bernd-edlinger reviewed (approved):
LGTM

dot-asm commented:
I get linking error on elderly Cygwin, there is no `mlock`...

levitte commented:
Hmmm, if I'm correctly informed, mlock() is a POSIX thing, and from my reading, it appeared in POSIX.1-2001, i.e. `_POSIX_VERSION == 200112L`.  That should make it a "safe" check of availability, wouldn't it?  Can you quickly check what POSIX version you have on that elderly Cygwin?

dot-asm commented:
`_POSIX_VERSION` is defined as 199009L. Which means that above mentioned will do. And it's only appropriate from formal viewpoint. I mean since function was added in X, it's only appropriate to check that X actually applies.

levitte committed:
Only implement secure malloc if _POSIX_VERSION allows

levitte commented:
Okie, rebased on fresher master, and one new commit that implements _POSIX_VERSION check on crypto/mem_sec.c

levitte commented on crypto/mem_sec.c:
```diff
@@ -20,7 +20,9 @@
-#if defined(OPENSSL_SYS_LINUX) || defined(OPENSSL_SYS_UNIX)
+/* e_os.h includes unistd.h, which defines _POSIX_VERSION */
+#if defined(OPENSSL_SYS_UNIX) \
+    && defined(_POSIX_VERSION) && _POSIX_VERSION >= 200112L
```
If anyone wonders why I removed `defined(OPENSSL_SYS_LINUX)`, it's because `OPENSSL_SYS_UNIX` is defined anyway on that platform...

dot-asm replied:
Well, formally speaking one can as well omit defined(OPENSSL_SYS_UNIX). As we already established it's rather a notion of POSIX compliance, and _POSIX_VERSION describes it better...

dot-asm replied:
But this merge request is probably not right place to discuss this kind of thing. Approving...

dot-asm commented:
Question now is if lack of mlock should actually translate to suppression of whole thing. I mean there are mmap and mprotect in older POSIX. And mlock is not actually required to succeed. I mean one can as well ret = 2 if `_POSIX_VERSION` is less than 200112L... But I suppose it's not inappropriate to do as suggested. Whole point with mem_sec.c is to prevent swapping (and core dumping). So if there is no mlock, it's really no point doing the rest...

dot-asm commented on engines/e_dasync.c:
```diff
@@ -25,7 +25,7 @@
 #include <openssl/modes.h>
 
-#if (defined(OPENSSL_SYS_UNIX) || defined(OPENSSL_SYS_CYGWIN)) && defined(OPENSSL_THREADS)
+#if defined(OPENSSL_SYS_UNIX) && defined(OPENSSL_THREADS)
```
On related *side* note, one can wonder if there are more dependencies on _POSIX_VERSION... Here or elsewhere... Subtle ones... Once again, *side* note!

levitte commented:
Cool!  Merged.

master: 33cebb9c0dd6115bc357b5097f7db82b64b10919 to e44c7d02ddac975ec6abff7901e77a0c37f9949d
1.1.0: df50dee414366bd609b1eadae8a0748e13a7226e to 2087582d0455111d6627a142668fdf0ebff1d89f

dot-asm commented:
_POSIX_VERSION strikes back ... on Solaris. By default, i.e. without additional flags, Solaris asserts rather old "POSIX identify", 199506L. Presumably to favour backward compatibility. This however does *not* affect visibility of mlock, in sense that mlock is declared even with such low _POSIX_VERSION. This applies even [at least!] to HP-UX, Tru64 and IRIX. I.e. they too assert older "POSIX identity", yet mlock is available/visible. In other words one can as well say that even though mlock was standardized in _POSIX_VERSION=200112L it was available from multiple vendors before that point. So one can argue that _POSIX_VERSION check can/should be complemented with vendor-specific checks. Just in case, trouble is that low _POSIX_VERSION translates to test failure.


 As "fun fact" in the context. Solaris' recommendation, according to `man standards`, is to `-D_POSIX_C_SOURCE` or `-D_XOPEN_SOURCE` to denote source compliance with more recent standard[s]. Fun part is that attempt to pass `-D_POSIX_C_SOURCE=200112L` to gcc results in `#error "... POSIX.1-2001 applications require the use of c99"`. I.e. it insist on complementing `-D_POSIX_C_SOURCE=200112L` with `-std=c99` (in case it's not default). For reference, passing `-D_POSIX_C_SOURCE=200112L` on other platforms mentioned above does *not* translate to corresponding `_POSIX_VERSION`, but to lower one.

bernd-edlinger commented:
GCC from GCC-5 on defaults to -std=gnu11 which is C11 with gnu extensions.

bernd-edlinger commented:
However ./config -D_POSIX_C_SOURCE=200112L causes build to fail for me:
```
gcc  -I. -Icrypto/include -Iinclude -DDSO_DLFCN -DHAVE_DLFCN_H -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -D_POSIX_C_SOURCE=200112L -DOPENSSLDIR="\"/usr/local/ssl\"" -DENGINESDIR="\"/usr/local/lib/engines-1.1\"" -Wall -O3 -pthread -m64 -DL_ENDIAN  -Wa,--noexecstack -c -fPIC -DOPENSSL_USE_NODELETE -MMD -MF crypto/async/arch/async_null.d.tmp -MT crypto/async/arch/async_null.o -c -o crypto/async/arch/async_null.o crypto/async/arch/async_null.c
In file included from /usr/include/ucontext.h:26,
                 from crypto/async/arch/../arch/async_posix.h:27,
                 from crypto/async/arch/../async_locl.h:30,
                 from crypto/async/arch/async_null.c:11:
/usr/include/x86_64-linux-gnu/sys/ucontext.h:137:5: error: unknown type name 'stack_t'
     stack_t uc_stack;
     ^~~~~~~
In file included from crypto/async/arch/../async_locl.h:30,
                 from crypto/async/arch/async_null.c:11:
crypto/async/arch/../arch/async_posix.h: In function 'async_fibre_swapcontext':
crypto/async/arch/../arch/async_posix.h:42:13: warning: implicit declaration of function '_longjmp'; did you mean 'longjmp'? [-Wimplicit-function-declaration]
             _longjmp(n->env, 1);
             ^~~~~~~~
             longjmp
```
